src/lib.rs

```diff
@@ -75,6 +77,89 @@ pub enum SymbolKind {
     Tls,
 }
 
+/// An iterator of the sections of an File
+pub struct SectionIterator<'a, 'b> {
+    inner: SectionIteratorInternal<'a, 'b>
+}
+
+impl<'a, 'b> fmt::Debug for SectionIterator<'a, 'b> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // It's painful to do much better than this
+        f.debug_struct("SectionIterator").finish()
+    }
+}
+
+// we wrap our enums in a struct so that they are kept private.
+enum SectionIteratorInternal<'a, 'b> {
+    MachO(Box<Iterator<Item=(mach::segment::Section, mach::segment::SectionData<'a>)> + 'b>),
+    Elf(slice::Iter<'a, elf::SectionHeader>, &'a elf::Elf<'a>, &'a [u8])
+}
+
+enum SectionInternal<'a> {
+    MachO((mach::segment::Section, mach::segment::SectionData<'a>)),
+    Elf(<slice::Iter<'a, elf::SectionHeader> as Iterator>::Item, &'a elf::Elf<'a>, &'a [u8])
+}
+
+/// A Section of a File
+pub struct Section<'a> {
+    inner: SectionInternal<'a>
+}
+
+impl<'a> fmt::Debug for Section<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // It's painful to do much better than this
+        f.debug_struct("Section")
+            .field("name", &self.name().unwrap_or("<invalid name>"))
+            .field("address", &self.address())
+            .field("size", &self.data().len())
+            .finish()
+    }
+}
+
+impl<'a> Section<'a> {
+    /// returns the address of the section
+    pub fn address(&self) -> u64 {
+        match &self.inner {
+            &SectionInternal::MachO(ref macho) => macho.0.addr,
+            &SectionInternal::Elf(ref elf, _, _) => elf.sh_addr
+        }
+    }
+
+    /// returns a reference to contents of the section
+    pub fn data(&self) -> &'a [u8] {
+        match &self.inner {
+            &SectionInternal::MachO(ref macho) => macho.1,
+            &SectionInternal::Elf(ref header, _, ref data) => &data[header.sh_offset as usize..][..header.sh_size as usize]
+        }
+    }
+
+    /// returns the name of the section
+    pub fn name(&self) -> error::Result<&str> {
```

Comment:
We need a better error handling strategy for this crate, but I'm not sure we should directly return goblin's `Result`. Would it be better to make this either `Option<&str>` or `Result<&str, &str>`?